### PR TITLE
sql/tidb: working ent integration tests

### PR DESCRIPTION
**Pull request by hedwigz**

### CODE REVIEW

File: internal/integration/go.mod [hunk 1]
```diff
@@ -5,8 +5,8 @@ go 1.17
 replace ariga.io/atlas => ../../
 
 require (
-	ariga.io/atlas v0.0.0
-	entgo.io/ent v0.9.2-0.20211116120827-6954eb3a182a
+	ariga.io/atlas v0.3.5-0.20220215131223-8043663b4223
```
Comment (a8m) [inline internal/integration/go.mod:8]:
```suggestion
	ariga.io/atlas v0.0.0
```
It's using `replace` above.

File: sql/mysql/tidb.go [hunk 2]
```diff
@@ -157,11 +159,40 @@ func (i *tinspect) patchSchema(ctx context.Context, s *schema.Schema) (*schema.S
...
+// e.g. []byte{4} -> b'100', []byte{2,1} -> b'1000000001'.
+// See: https://github.com/pingcap/tidb/issues/32655.
+func bytesToBitLiteral(b []byte) string {
+	bytes := make([]byte, 8)
+	for i := 0; i < len(b); i++ {
+		bytes[8-len(b)+i] = b[i]
+	}
+	val := binary.BigEndian.Uint64(bytes)
+	return fmt.Sprintf("b'%b'", val)
+}
```
Comment (a8m) [inline sql/mysql/tidb.go:193]: Please rebase on master.

File: sql/mysql/tidb.go [hunk 3]
```diff
@@ -231,13 +262,40 @@ var reColl = regexp.MustCompile(`(?i)CHARSET\s*=\s*(\w+)\s*COLLATE\s*=\s*(\w+)`)
...
+		return nil
+	}
+	v, err := strconv.ParseInt(matches[1], 10, 64)
+	if err != nil {
+		return err
+	}
+	ai.V = v
+	schema.ReplaceOrAppend(&t.Attrs, ai)
+	return nil
+}
```
Comment (a8m) [inline sql/mysql/tidb.go:301]: Please see https://github.com/ariga/atlas/blob/master/sql/mysql/inspect.go#L483-L508

Comment (hedwigz) [reply sql/mysql/tidb.go:301]: I saw that but I didn't understand why it needs `putShow` which isn't used by me

File: go.mod [hunk 4]
```diff
@@ -30,7 +29,10 @@ require (
 	github.com/chzyer/readline v0.0.0-20180603132655-2972be24d48e // indirect
 	github.com/davecgh/go-spew v1.1.1 // indirect
 	github.com/inconshreveable/mousetrap v1.0.0 // indirect
+	github.com/kr/pretty v0.2.1 // indirect
```
Comment (a8m) [inline go.mod:32]: ?

File: internal/integration/go.mod [hunk 5]
```diff
@@ -5,8 +5,8 @@ go 1.17
 replace ariga.io/atlas => ../../
 
 require (
-	ariga.io/atlas v0.0.0
-	entgo.io/ent v0.9.2-0.20211116120827-6954eb3a182a
+	ariga.io/atlas v0.3.7-0.20220303204946-787354f533c3
```
Comment (a8m) [inline internal/integration/go.mod:8]:
```suggestion
	ariga.io/atlas v0.0.0
```

We use replace.

File: internal/integration/tidb_test.go [hunk 6]
```diff
@@ -5,13 +5,18 @@
 package integration
 
 import (
+	"ariga.io/atlas/sql/mysql"
+	"ariga.io/atlas/sql/schema"
+
 	"context"
```
Comment (a8m) [inline internal/integration/tidb_test.go:11]: goimports. merge the two blocks

File: sql/mysql/tidb.go [hunk 7]
```diff
@@ -258,13 +264,40 @@ var reColl = regexp.MustCompile(`(?i)CHARSET\s*=\s*(\w+)\s*COLLATE\s*=\s*(\w+)`)
...
-		return fmt.Errorf("missing COLLATE and/or CHARSET information on CREATE TABLE statment for %q", t.Name)
+		return fmt.Errorf("missing COLLATE and/or CHARSET information on CREATE TABLE statement for %q", t.Name)
 	}
 	t.SetCharset(matches[1])
 	t.SetCollation(matches[2])
 	return nil
 }
+
+// e.g AUTO_INCREMENT=1234
+var reAI = regexp.MustCompile(`(?i)\s*AUTO_INCREMENT\s*=\s*(\d+)\s*`)
```
Comment (a8m) [inline sql/mysql/tidb.go:279]: There's a similar regexp in this package. `reAutoinc`

File: internal/integration/tidb_test.go [hunk 8]
```diff
@@ -479,6 +484,173 @@ create table atlas_defaults
...
+	empty := `
+schema "test" {
+}
+`
+	tidbRun(t, func(t *myTest) {
+		testHCLIntegration(t, full, empty)
+	})
+}
+
+func TestTiDB_Ent(t *testing.T) {
```
Comment (a8m) [inline internal/integration/tidb_test.go:621]: Can't you use the existing ent-integration test?

Comment (hedwigz) [reply internal/integration/tidb_test.go]: The new TiDB ent integration test uses atlas as a migration engine.  
I also added the old engine tests, using `migrate.WithForeignKeys(false)` to make it work.

File: go.mod (hunk 4 above)
Comment (hedwigz) [reply go.mod:32]: https://github.com/ent/ent/pull/2337/files#r808138254  
flashback?

File: internal/integration/integration_test.go [hunk 9]
```diff
@@ -4,6 +4,7 @@ import (
 	"bytes"
 	"context"
 	"database/sql"
+	entschema "entgo.io/ent/dialect/sql/schema"
```
Comment (a8m) [inline internal/integration/integration_test.go:7]: move to line 21

File: internal/integration/tidb_test.go [hunk 10]
```diff
@@ -5,13 +5,18 @@
 package integration
 
 import (
+	"ariga.io/atlas/sql/mysql"
+	"ariga.io/atlas/sql/schema"
 	"context"
 	"database/sql"
+	entschema "entgo.io/ent/dialect/sql/schema"
```
Comment (a8m) [inline internal/integration/tidb_test.go:12]: ditto

**Review by a8m — APPROVED** [review]
Looks good! Please, squash and merge, and fix commit message

**Pull request merged**